pkg/querier/iterators/chunk_iterator.go

```diff
@@ -86,17 +86,13 @@ func (i *chunkIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
 		return i.cachedTime, i.cachedFloatHistogram
 	}
 
-	var (
-		t  int64
-		fh *histogram.FloatHistogram
-	)
-
+	var t int64
 	if i.valType == chunkenc.ValHistogram {
 		var h *histogram.Histogram
-		t, h = i.AtHistogram()
-		fh = h.ToFloat(nil)
+		t, h = i.AtHistogram(i.cachedHistogram)
+		fh = h.ToFloat(fh)
```

Comment:
I don't think we need `h` here. If the chunk only has float histograms, i.cachedHistogram would start from nil, and we don't keep it in this line, so it will always be nil, I'd suggest removing `h` and:


```suggestion
		t, i.cachedHistogram = i.AtHistogram(i.cachedHistogram)
		fh = i.cachedHistogram.ToFloat(fh)
```

Reply:
By the way, the weird thing here is that the integer histogram raw chunks can execute direct on [AtFloatHistogram](https://github.com/prometheus/prometheus/blob/419dd265ccd198fbf312659b3671dac2ecaad510/tsdb/chunkenc/histogram.go#L821) , but interface doesn't tell if that's possible or not, unless we do a type assertion maybe. But let's have that discussion later, a better solution would be to require AtFloatHistogram to work even if type is ValHistogram...

Reply:
I agree with the proposed change, and I have updated the PR.